src/v/cloud_topics/frontend_reader/reader.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "cloud_topics/level_one/common/abstract_io.h"
+#include "cloud_topics/level_one/metastore/metastore.h"
+#include "cloud_topics/level_zero/stm/ctp_stm_api.h"
+#include "cloud_topics/log_reader_config.h"
+#include "model/record_batch_reader.h"
+
+namespace cluster {
+class partition;
+}
+
+namespace cloud_topics {
+class data_plane_api;
+
+namespace l1 {
+class frontend;
+}
+
+/*
+ * This class implements a reader for a cloud topic. It reads from both L0 and
+ * L1.
+ *
+ * The reader determines at initialization whether to start reading from L1 or
+ * L0 based on the LRO, and if the L1 reader is exhausted but the reader expects
+ * more batches, the reader transitions to reading from L0.
+ */
+class cloud_topics_log_reader_impl : public model::record_batch_reader::impl {
```

Comment:
Do we actually need joining reader?
We can just create L1 reader or L0 reader depending on the starting offset. When the consumer will read up to the end of the L1 it will keep reading and will get L0 reader.